agent/runner/runner.go

```diff
@@ -0,0 +1,270 @@
+	select {
+	case r.jobs <- job:
+		return nil
+	default:
+		return errors.New("jobs queue overflowed")
+	}
+}
+
+// JobsMessages returns channel with Jobs messages.
+func (r *Runner) JobsMessages() <-chan agentpb.AgentResponsePayload {
+	return r.jobsMessages
+}
+
+// ActionsResults return chanel with Actions results payload.
+func (r *Runner) ActionsResults() <-chan agentpb.AgentRequestPayload {
+	return r.actionsMessages
+}
+
+// Stop stops running Action or Job.
+func (r *Runner) Stop(id string) {
+	r.rw.RLock()
+	defer r.rw.RUnlock()
+
+	// Job removes itself from rCancel map. So here we only invoke cancel.
+	if cancel, ok := r.rCancel[id]; ok {
+		cancel()
+	}
+}
+
+// IsRunning returns true if Action or Job with given ID still running.
+func (r *Runner) IsRunning(id string) bool {
+	r.rw.RLock()
+	defer r.rw.RUnlock()
+	_, ok := r.rCancel[id]
+
+	return ok
+}
+
+func (r *Runner) handleJob(ctx context.Context, job jobs.Job) {
+	jobID, jobType := job.ID(), job.Type()
+	l := r.l.WithFields(logrus.Fields{"id": jobID, "type": jobType})
+
+	if err := r.sem.Acquire(ctx, 1); err != nil {
+		l.Errorf("Failed to acquire token for a job: %v", err)
+		r.sendJobsMessage(&agentpb.JobResult{
+			JobId:     job.ID(),
+			Timestamp: timestamppb.Now(),
+			Result: &agentpb.JobResult_Error_{
+				Error: &agentpb.JobResult_Error{
+					Message: err.Error(),
+				},
+			},
+		})
+		return
+	}
+
+	var nCtx context.Context
+	var cancel context.CancelFunc
+	if timeout := job.Timeout(); timeout != 0 {
+		nCtx, cancel = context.WithTimeout(ctx, timeout)
+	} else {
+		nCtx, cancel = context.WithCancel(ctx)
+	}
+	r.addCancel(jobID, cancel)
+
+	r.wg.Add(1)
+	run := func(ctx context.Context) {
+		l.Infof("Job started.")
+
+		defer func(start time.Time) {
+			l.WithField("duration", time.Since(start).String()).Info("Job finished.")
+		}(time.Now())
+
+		defer r.sem.Release(1)
+		defer r.wg.Done()
+		defer cancel()
+		defer r.removeCancel(jobID)
+
+		err := job.Run(ctx, r.sendJobsMessage)
+		if err != nil {
+			r.sendJobsMessage(&agentpb.JobResult{
+				JobId:     job.ID(),
+				Timestamp: timestamppb.Now(),
+				Result: &agentpb.JobResult_Error_{
+					Error: &agentpb.JobResult_Error{
+						Message: err.Error(),
+					},
+				},
+			})
+			l.Warnf("Job terminated with error: %+v", err)
+		}
+	}
+
+	go pprof.Do(nCtx, pprof.Labels("jobID", jobID, "type", string(jobType)), run)
+}
+
+func (r *Runner) handleAction(ctx context.Context, action actions.Action) {
+	actionID, actionType := action.ID(), action.Type()
+	l := r.l.WithFields(logrus.Fields{"id": actionID, "type": actionType})
+
+	if err := r.sem.Acquire(ctx, 1); err != nil {
+		l.Errorf("Failed to acquire token for an action: %v", err)
+		r.sendActionsMessage(&agentpb.ActionResultRequest{
+			ActionId: actionID,
+			Done:     true,
+			Error:    err.Error(),
+		})
+		return
+	}
+
+	var timeout time.Duration
+	if timeout = action.Timeout(); timeout == 0 {
+		timeout = defaultActionTimeout
+	}
```

Comment:
we can get rid of this just by setting default value in the place above

Reply:
I don't get the idea. 🤔

Reply:
in the place we discussed above we set `0` as the value and then we update it here to a default value

Reply:
Aha! Well, this is logic of the runner. It doesn't let to run actions without timeout (unlike jobs). So if we will set timeout in `client.go` that means that runner somehow should process actions with zero timeout.

Reply:
it will fail the moment it started, but ok. let's proceed.